834 targeted rediscretization

IvarStefansson opened the pull request:
## Proposed changes

Facilitates targeted rediscretization. Closes #834. Do we need a specific test for this functionality?

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR)?

jwboth commented:
Should also close #841 (I believe it is a subset of #834 which I oversaw...)

keileg commented on src/porepy/models/solution_strategy.py:
```diff
@@ -113,6 +113,7 @@ def __init__(self, params: Optional[dict] = None):
         The tuple contains the sparse matrix and the right hand side residual vector.
 
         """
+        self._nonlinear_discretizations: list[pp.ad._ad_utils.MergedOperator] = []
```
Documentation please.

keileg replied:
You're right, I did not consider the property part.

keileg commented on src/porepy/models/mass_and_energy_balance.py:
```diff
@@ -89,7 +89,13 @@ class SolutionStrategyFluidMassAndEnergy(
     """
 
-    pass
+    # def set_nonlinear_discretizations(self) -> None:
```
Can we delete this?

IvarStefansson replied:
Yes, sorry

keileg commented on src/porepy/models/poromechanics.py:
```diff
@@ -261,6 +267,18 @@ def _is_nonlinear_problem(self) -> bool:
+        # Nonlinear discretizations for the fluid mass balance subproblem. The momentum
+        # balance does not have any.
+        super().set_nonlinear_discretizations()
+        # Aperture changes render permeability variable. This requires a re-discretization
+        # of the diffusive flux in subdomains where the aperture changes.
```
```suggestion
        # Aperture changes render permeability variable. This requires a 
        # re-discretization # of the diffusive flux in subdomains where the aperture
        # changes.
```

IvarStefansson replied:
What's going on here?

keileg replied:
I think you were above the 88 lines. If not, disregard.

keileg commented on src/porepy/models/thermoporomechanics.py:
```diff
@@ -163,6 +177,23 @@ def set_discretization_parameters(self) -> None:
+        # Super calls method in mass and energy balance. Momentum balance has no
+        # nonlinear discretizations.
+        super().set_nonlinear_discretizations()
+        # Aperture changes render permeability variable. This requires a re-discretization
+        # of the diffusive flux in subdomains where the aperture changes.
```
```suggestion
        # Aperture changes render permeability variable. This requires a 
        # re-discretization # of the diffusive flux in subdomains where the aperture
        # changes.
```

keileg commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -1658,12 +1664,15 @@ def advective_flux(
             Operator representing the advective flux.
 
         """
+        # The upwind discretization is non-linear (i.e. depends on solution state).
+        # Append to list of discretizations to be updated at each nonlinear iteration.
+        # self.add_nonlinear_discretization(discr)
+
```
Should this be deleted? Intuitively, it seems the solution strategy is the better option.

keileg commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -1702,6 +1711,9 @@ def interface_advective_flux(
             Operator representing the advective flux on the interfaces.
 
         """
+        # The upwind discretization is non-linear (i.e. depends on solution state).
+        # Append to list of discretizations to be updated at each nonlinear iteration.
+        # self.add_nonlinear_discretization(discr)
```
Delete this?

keileg commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -1627,6 +1627,12 @@ class AdvectiveFlux:
     """Well flux variables on interfaces. Normally defined in a mixin instance of
     :class:`~porepy.models.fluid_mass_balance.VariablesSinglePhaseFlow`.
+    """
+    add_nonlinear_discretization: Callable[[pp.ad.Discretization], None]
+    """Add a discretization to the model's list of discretizations to be updated at each
+    nonlinear iteration. Normally defined in a mixin instance of
+    :class:`~porepy.models.solution_strategy.SolutionStrategy`.
+
     """
```
Delete this if the parts commented out below are deleted.

keileg commented on src/porepy/models/solution_strategy.py:
```diff
@@ -213,6 +215,49 @@ def discretize(self) -> None:
+    def set_nonlinear_discretizations(self) -> None:
+        """Set the list of nonlinear discretizations.
+
+        This method is called before the discretization is performed. It is
+        intended to be used to set the list of nonlinear discretizations.
```
Please fix line length. I did not try my luck to hit the 88 characters..

keileg reviewed (commented):
Nice work. I also agree with the design choices made.

From GH actions, it may seem the runtime for the integration test is cut by a minute - if so, that will be most useful.

Testing: I am inclined to say that this is tested along with the integration tests of the models. The two things that can go wrong (I think) is that we either discretize everything (as done before), or nothing. Both seem to have low likelihood of occurence, so I'm not sure it is worth the extra time. What do you think?

IvarStefansson commented:
> Nice work. I also agree with the design choices made.
> 
> From GH actions, it may seem the runtime for the integration test is cut by a minute - if so, that will be most useful.
> 
> Testing: I am inclined to say that this is tested along with the integration tests of the models. The two things that can go wrong (I think) is that we either discretize everything (as done before), or nothing. Both seem to have low likelihood of occurence, so I'm not sure it is worth the extra time. What do you think?

There's a third pitfall which I guess is more likely and harder to detect: We forget adding one or more discretizations to the list, and thus end up not discretizing those. Full coverage is costly, but we could do one for thermoporomechanics, which would cover quite a bit.

keileg commented on tests/integration/models/test_rediscretization.py:
```diff
@@ -0,0 +1,129 @@
+        if self._nonlinear_iteration > 1:
+            return 0.0, True, False
+        else:
+            return super().check_convergence(*args, **kwargs)
```
Super without inheritance?

keileg commented:
Thanks for adding the test. I added some comments which will make it easier (for me) to revisit the test if I ever have to, and changed the matrix comparison to work on dense matrices - that should be very cheap and leave no room for differing sparsity patterns.

IvarStefansson commented:
> Thanks for adding the test. I added some comments which will make it easier (for me) to revisit the test if I ever have to, and changed the matrix comparison to work on dense matrices - that should be very cheap and leave no room for differing sparsity patterns.

Does that mean you're ready to approve the PR?

keileg reviewed (approved):
Sorry, I though I already had approved.
